datanode/dehistory/initialise.go

```diff
@@ -215,33 +224,43 @@ func GetMostRecentHistorySegmentFromPeer(ctx context.Context, ip string, datanod
 		return nil, fmt.Errorf("failed to get most recent history segment:%w", err)
 	}
 
-	return resp.GetSegment(), nil
+	return resp, nil
 }
 
 // TODO this needs some thought as to the best strategy to select the root segment to avoid spoofing.
-func SelectRootSegment(peerToSegment map[string]*v2.HistorySegment) *v2.HistorySegment {
-	segmentsList := make([]*v2.HistorySegment, 0, len(peerToSegment))
+func SelectMostRecentHistorySegmentResponse(peerToSegment map[string]*v2.GetMostRecentDeHistorySegmentResponse, swarmKey string) *PeerResponse {
+	segmentsList := make([]PeerResponse, 0, len(peerToSegment))
 
-	for _, segment := range peerToSegment {
-		segmentsList = append(segmentsList, segment)
+	for peer, segment := range peerToSegment {
+		segmentsList = append(segmentsList, PeerResponse{peer, segment})
 	}
 
 	// Sort history latest first
 	sort.Slice(segmentsList, func(i, j int) bool {
-		return segmentsList[i].ToHeight > segmentsList[j].ToHeight
+		return segmentsList[i].Response.Segment.ToHeight > segmentsList[j].Response.Segment.ToHeight
 	})
 
+	// Filter out segments without matching swarm key
+	var filteredSegments []PeerResponse
+	for _, segment := range segmentsList {
+		if segment.Response.SwarmKey == swarmKey {
+			filteredSegments = append(filteredSegments, segment)
+		}
+	}
+
 	// Filter out segments with toHeight < highest toHeight
-	var filteredSegments []*v2.HistorySegment
+	segmentsList = filteredSegments
+	filteredSegments = nil
 	for _, segment := range segmentsList {
-		if segment.ToHeight == segmentsList[0].ToHeight {
+		if segment.Response.Segment.ToHeight == segmentsList[0].Response.Segment.ToHeight {
```

Comment:
Comment seems invalid here, aren't we instead filtering out segment where ToHeight is < highest && ToHeight > highest?
Is that the expected behaviour?

Reply:
yeah, the method was unclear, refactored it